### PR TITLE
docs: add epm desgin document

### PR DESCRIPTION
**Pull request by YangLiang3**
Fixes: #363
Signed-off-by: Liang Yang <liang3.yang@intel.com>

### CODE REVIEW

File: docs/design/epm/design.md [hunk 1]
```diff
@@ -0,0 +1,206 @@
+# Overview
+
+This document introduces EPM fundamental and design. EPM is a module optimized enclave boot time. It is designed as two parts, one is as OCI bundles cache management, the other one is as enclave cache pool management. This documentation will focus on enclave cache pool management currently, briefly call it as EPM as below.
```
Comment (hustliyilin) [inline docs/design/epm/design.md:3]: It's better to add the URL of `epm` when it first appears.

Comment (hustliyilin) [inline docs/design/epm/design.md:3]: change `one is as OCI bundles cache management, the other one is as enclave cache pool management` to `one is OCI bundles cache management, the other one is enclave cache pool management`

File: docs/design/epm/design.md [hunk 2]
```diff
@@ -0,0 +1,206 @@
+# Overview
+
+This document introduces EPM fundamental and design. EPM is a module optimized enclave boot time. It is designed as two parts, one is as OCI bundles cache management, the other one is as enclave cache pool management. This documentation will focus on enclave cache pool management currently, briefly call it as EPM as below. 
+
+The EPM is used to cache enclaves which can store various enclave runtime, e.g. skeleton, Graphene (under plan), Occlum(under development). Current implementation is based on skeleton using sgx1 not supporting to extend extra pages once the enclave initialization is done. Inclavare-containers provides skeleton as an enclave runtime reference. All the introduction to EPM is based on skeleton.
```
Comment (hustliyilin) [inline docs/design/epm/design.md:5]: Add the URL of `skeleton`、`Graphene`, and `Occlum` when they firstly appear.

File: docs/design/epm/design.md [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+EPM includes EPM service and EPM client sorted by function.
+
+- EPM service is used to store the enclave info and as gRPC server implementing EPM interface. There are two communicating channels between EPM service and EPM client. Channel one is used by gRPC as function calling. Channel two is used by transferring enclave file descriptor between both processes. EPM service will listen on epm.sock in unix socket. It will receive the gRPC calling from EPM client.
+- The EPM client can collect the enclave’s info by procfs and transfer skeleton enclave info to enclave pool cache in EPM service. Skeleton enclave info includes enclave base address, enclave descriptor file, enclave memory layout, permissions. 
+
+In container pal API will be used to manage enclave runtime lifecycle. Pal API V3 provides EPM reference implementation.
```
Comment (hustliyilin) [inline docs/design/epm/design.md:12]: change `In container pal API` to `Enclave runtime pal API`

File: docs/design/epm/design.md [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+one is pre-storing, the other is final-storing. The enclave file descriptor just can be transferred with the same process who is opening the enclave. This is why to use two stages of storing. In pre-storing state, the enclave info is not valid to be used, it just saves the fd to EPM service by unix domin socket with sendmsg/recvmsg. Once the enclave is released by pal_destroy in container, the enclave info except enclave fd will be transferred by gRPC interface. The pal_destroy will call gRPC interface to sync with the EPM service, until now the enclave info will be available for other containers. Once other container is started, pal_init will try to query and get an enclave from enclave pool by gRPC interface. If the enclave pool is not empty, the enclave info will be responded to EPM client. After mmap with parameter enclave file descriptor and enclave map address, the enclave will be valid to container.
+
+## EPM Service implementation
+
+### EPM service process
+
+EPM is started as a process on host. It’s based on gRPC inter-process communicating. EPM service is started by command line: sudo epm. It will create and listen on /var/run/epm/epm.sock as following:
+
+```code
```
Comment (hustliyilin) [inline docs/design/epm/design.md:32]: change to ```Go

File: docs/design/epm/design.md [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+// listen and serve
+lis, err := net.Listen("unix", cfg.GRPC.Address)
+s.Serve(lis)
+```
+
+### Enclave and cache pool structure
+
+1. Enclave proto definition:
+
+```code
```
Comment (hustliyilin) [inline docs/design/epm/design.md:49]: change to ```Protobuf

File: docs/design/epm/design.md [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+        bool execute = 3;
+        bool share = 4;
+        bool private = 5;
+}
+```
+
+2. Cache and Cache Pool:
+Cache represents the metadata of a cache managed by enclave pool.
+
+```code
```
Comment (hustliyilin) [inline docs/design/epm/design.md:72]: ditto

File: docs/design/epm/design.md [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+    int64 size = 6; // Size represents the size in bytes of the cache
+    int64 created = 7; // Created represents the creation time of the cache which is the number of seconds elapsed since January 1, 1970 UTC
+    google.protobuf.Any options = 8; // Options is an optional field which can extend any type of data structure
+}
+```
+
+3. EnclavePoolManager
+EnclavePoolManager represents an enclave pool.
+
+```code
```
Comment (hustliyilin) [inline docs/design/epm/design.md:88]: ditto

File: docs/design/epm/design.md [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+    rpc ListCache(ListCacheRequest) returns (ListCacheResponse) {}
+    // DeleteCache represents delete the specified cached data and remove the corresponding cache metadata
+    rpc DeleteCache(DeleteCacheRequest) returns (DeleteCacheResponse) {}
+    // LoadCache represents load the specified cache data to work directory
+    rpc LoadCache(LoadCacheRequest) returns (LoadCacheResponse) {}
+}
+```
+
+Enclave info will be stored into the map variant temporarily after pal_init.
+var EnclavePoolPreStore map[string]*v1alpha1.Enclave
```
Comment (hustliyilin) [inline docs/design/epm/design.md:108]: add ```go

File: docs/design/epm/design.md [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+Enclave info will be stored into the map variant finally after pal_destroy.
+var EnclavePoolStore map[int]*v1alpha1.Enclave
```
Comment (hustliyilin) [inline docs/design/epm/design.md:110]: add ```go

File: docs/design/epm/design.md [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+each enclave manager will implement the following interface, currently we just have one type of enclave runtime, we implement it in enclave-cache.go
+
+4. EnclavePool interface
+
+EnclavePool represents one type of enclave pool, each kind of enclave pool need implement the EnclavePool interface.
+
+```code
```
Comment (hustliyilin) [inline docs/design/epm/design.md:117]: ditto

File: docs/design/epm/design.md [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+        // GetPoolType gets the pool type of current pool
+        GetPoolType() string
+}
+```
+
+5. DefaultEnclavePool
+
+DefaultEnclavePool is the default implementation of EnclavePool
+
+```code
```
Comment (hustliyilin) [inline docs/design/epm/design.md:140]: ditto

File: docs/design/epm/design.md [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+        Enclaveinfo   map[int]*v1alpha1.Enclave
+        CacheMetadata *cache_metadata.Metadata
+}
+```
+
+### Enclave cache pool operation
+
+We just need implement two interfaces:
+
+func (d *EnclaveCacheManager) GetCache(ID string) (*v1alpha1.Cache, error)
```
Comment (hustliyilin) [inline docs/design/epm/design.md:153]: add ```go

File: docs/design/epm/design.md [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+
+This function will marshal enclave info in cache pool, and then sendmsg enclave file descriptor to the receiving process init-runelet in container. By sendmsg, the enclave file descriptor will be dup to init-runelet process in container which will create and map the enclave by enclave info.
+
+![getcache](getcache.png)
+
+func (d *EnclaveCacheManager) SaveCache(sourcePath string, cache *v1alpha1.Cache) error
```
Comment (hustliyilin) [inline docs/design/epm/design.md:159]: ditto

File: docs/design/epm/design.md [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+
+SaveCache represents enclave info will be saved into EnclavePoolStore, the control flow is as below:
+
+![savecache](savecache.png)
+
+## EPM client implementation
+
+![epmclient](epmclient.png)
+
+Save enclave from EPM client to EPM service:
```
Comment (hustliyilin) [inline docs/design/epm/design.md:169]: change to `###Save enclave from EPM client to EPM service`

File: docs/design/epm/design.md [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+1. analyze /proc/self/maps, get device /dev/sgx/enclave’s mapping info(address, size, permission, flag)
+2. query /proc/self/fd, get device /dev/sgx/enclave’s fd
+3. grpc those info above to EPM service, but enclave fd is sent by sendmsg in unix socket.
+Get enclave from EPM service to EPM client:
```
Comment (hustliyilin) [inline docs/design/epm/design.md:173]: change to `### Get enclave from EPM service to EPM client`

File: docs/design/epm/design.md [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+1. Calling GetCache in pal_init which will be introduced in 2.4 chapter, The GetCache function will go into EPM service by gRPC. 
+2. cacheResp will be responded with cache info from EPM service to EPM client. Enclave fd will be returned by sendmsg/recvmsg. Enclave info will be composed of cacheResp and enclave fd.
+
+## Enclave Pal API
```
Comment (hustliyilin) [inline docs/design/epm/design.md:177]: It's better to quote the [Enclave Runtime PAL API Specification](https://github.com/alibaba/inclavare-containers/blob/master/rune/libenclave/internal/runtime/pal/spec.md#enclave-runtime-pal-api-specification) directly.

File: docs/design/epm/design.md (hunk 2 above)
Comment (haosanzi) [reply docs/design/epm/design.md:5]: using `sgx1` or `sgx2`?

File: docs/design/epm/design.md [hunk 17]
```diff
@@ -0,0 +1,206 @@
...
+```code
+message Enclave {
+        int64 fd = 1; //Enclave file descriptor in current process
+        int64 nr = 2; //Enclave memory segment numbers
+        repeated Enclavelayout layout = 3;//Enclave memory layout
+}
+message Enclavelayout {
+        uint64 addr = 1; //Current enclave segment starting address
+        uint64 size = 2;//Current enclave segment size
+        EnclavePerms prot = 3;//Current enclave segment’s permission
```
Comment (haosanzi) [inline docs/design/epm/design.md:58]: Please unity the format, use `;//` or `; //`

**Review by haosanzi — APPROVED** [review]

**Review by haosanzi — COMMENTED** [review]
some code or function in the sentence， it is better to use `function` to distinguish

File: docs/design/epm/design.md (hunk 2 above)
Comment (YangLiang3) [reply docs/design/epm/design.md:5]: sgx1

File: docs/design/epm/design.md [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+1. Analyze /proc/self/maps, get device /dev/sgx/enclave’s mapping info(address, size, permission, flag)
+2. Query /proc/self/fd, get device /dev/sgx/enclave’s fd
+3. GRPC those info above to EPM service, but enclave fd is sent by sendmsg in unix socket.
+
+### Get enclave from EPM service to EPM client:
+
+1. Calling GetCache in pal_init, The GetCache function will go into EPM service by gRPC. 
+2. cacheResp will be responded with cache info from EPM service to EPM client. Enclave fd will be returned by sendmsg/recvmsg. Enclave info will be composed of cacheResp and enclave fd.
+
+## Enclave Pal API
```
Comment (hustliyilin) [inline docs/design/epm/design.md:192]: Change to `Enclave Runtime PAL API`

**Review by hustliyilin — COMMENTED** [review]
You also need to change [epm README.md](https://github.com/alibaba/inclavare-containers/blob/master/epm/README.md) to quote the `epm/design.md` you committed

**Review by hustliyilin — APPROVED** [review]

**Review by haosanzi — APPROVED** [review]

**Review by jiazhiguang — APPROVED** [review]

**Pull request merged**